[DNNL]Add support for int8Conv

lisa0314 opened the pull request:
Fix [intel/webml-polyfill#1166](https://github.com/intel/webml-polyfill/issues/1166)
@huningxin , please help to review. Thanks!

sunlin-link commented:
@lisa0314 @fujunwei Please help to review

fujunwei commented on services/ml/compilation_impl_mac.mm:
```diff
@@ -1065,15 +1065,13 @@ void ComputeBNNSOffsetForImplicitPadding(
         std::make_unique<ExecutionImplMac>(this, std::move(memory_handle));
-    if (!impl->IsValid()) {
-      std::move(callback).Run(mojom::BAD_DATA, std::move(init_params));
-      return;
```
Does it crash in binding interface  or js object, could you please explain the root cause in detail, thanks.

sunlin-link replied:
`init_params->execution` should not be empty, or it will cause crash.

fujunwei replied:
Where does it crash? do you have stack, the [L51]( https://github.com/otcshare/chromium-src/blob/webml/third_party/blink/renderer/modules/ml/execution.cpp#L51) uses the variable, but i think it doesn't reach.

sunlin-link replied:
The crash is inside std::move. By the case https://github.com/intel/webml-polyfill/issues/58

fujunwei replied:
So the root cause of it is we can't pass `init_params` which is invalid, you can fix it with nullptr

if (!impl->IsValid()) {
    std::move(callback).Run(mojom::BAD_DATA, nullptr);
}

sunlin-link replied:
Although the construction of execution is incomplete, `init_params` contains other variables like `inputs`, `outpus`, `memory` with no problem. These things maybe useful for further usage. Furthermore, the user may need to know why the construction of execution is failed. Currently, if the result is `BAD_DATA`, the caller does nothing and just throw a DOM exeception. But it doesn't mean `init_params` won't be used in the future.

fujunwei replied:
I think it should be added when needed,  it's not necessary in [this case](https://github.com/otcshare/chromium-src/blob/webml/third_party/blink/renderer/modules/ml/compilation.cpp#L115).

huningxin replied:
If there are no existing usages of `init_params` when construction of execution fails, I prefer to passing `nullptr`.

sunlin-link commented:
Fixed

fujunwei commented:
LGTM.

huningxin commented:
Thanks for the PR and review!

sunlin-link commented:
Fixes:
https://github.com/intel/webml-polyfill/issues/164
https://github.com/intel/webml-polyfill/issues/173

which is regression by https://github.com/otcshare/chromium-src/pull/11

lisa0314 committed:
[DNNL]Add support for int8Conv

huningxin commented:
Thanks @lisa0314 , did you happen to test against new quant8 test cases of https://github.com/intel/webml-polyfill/pull/1162 and https://github.com/intel/webml-polyfill/pull/1162.

lisa0314 commented:
The current PR only support symm(zero_point=0) and I have created a new [issue](https://github.com/intel/webml-polyfill/issues/1193) to follow the implementation of asymm in the future.
Regression test results:
https://brucedai.github.io/webnnt/test/index-local.html?prefer=fast
passes: 436 failures: 462(baseline: "pass": 436, "fail": 462)

For intel/webml-polyfill#1162:
Pass the test case when zero_point = 0 **(8/40)** 

@huningxin Please help to review agian. Thanks!

huningxin commented:
Thanks @lisa0314 , could you please share the results of TENSOR_QUANT8_ASYMM_SIGNED and TENSOR_QUANT8_SYMM_PER_CHANNEL by https://github.com/intel/webml-polyfill/pull/1180 and https://github.com/intel/webml-polyfill/pull/1162? It include 40 test cases. Thanks.

lisa0314 commented:
@huningxin I have update the results of quant cts test and it pass 8 of 40 totally. Since DNNL can't support weights of u8 and zeropoint,  most cases fail. The past cts test cases are listed as below:

- Conv2d quant8 signed example-1

- Conv2d quant8 signed example-2

- Conv2d v1_2 example-1

- Conv2d v1_2 example-2

- Conv2d v1_2 example-5

- Conv2d v1_2 example-6

- Depthwise conv2d per channel example-1

- Depthwise conv2d per channel example-2 

Since ohwi format is not supported by s8->s8 reorder by DNNL currently -[ issue](https://github.com/oneapi-src/oneDNN/issues/691) , I just use a workaround to reorder ohwi to oihw as the first step. After this issue is fixed, I will delete related code.

huningxin commented on services/ml/common.cc:
```diff
@@ -27,6 +27,10 @@ uint32_t GetRequiredSize(int32_t type,
     return product(dimensions) * sizeof(int32_t);
   } else if (type == mojom::TENSOR_QUANT8_ASYMM) {
     return product(dimensions) * sizeof(int8_t);
```
Should be `uint8_t`. This is not introduced in this PR, but it would be good to fix that.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -450,6 +455,7 @@ int32_t CompilationDelegateDnnl::AddReorder(const std::string& input_name,
     LOG(ERROR) << "[DNNL] failed to get memory descriptor " << status;
     return mojom::OP_FAILED;
   }
+
```
This new line is not needed.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -957,6 +987,56 @@ int32_t CompilationDelegateDnnl::AddConvolution(
   }
+
+  // workaround for support s8 weights reorder from ohwi
+  // since DNNL can't support by now
```
Could you please put a link to the tracking issue of DNNL?

huningxin commented on services/ml/compilation_impl.h:
```diff
@@ -32,6 +32,13 @@ struct ElementWiseParams {
+  int32_t bias_type;
+  int32_t output_mask;
+  uint32_t output_scales_count;
+  float* output_scales;
```
If they are only used by dnnl, it would be better to move to dnnl specific structures.

huningxin commented on services/ml/model_impl.cc:
```diff
@@ -16,8 +16,6 @@ ModelImpl::~ModelImpl() = default;
 void ModelImpl::Finish(mojom::ModelInfoPtr model_info,
                        FinishCallback callback) {
-  DLOG(INFO) << "ModelImpl::Finish";
-
```
Any reasons to delete this?

huningxin commented on services/ml/neural_network_impl.cc:
```diff
@@ -23,7 +23,6 @@ NeuralNetworkImpl::NeuralNetworkImpl() = default;
 NeuralNetworkImpl::~NeuralNetworkImpl() = default;
 
 void NeuralNetworkImpl::CreateModel(CreateModelCallback callback) {
-  DLOG(INFO) << "Create Model";
```
Any reasons to delete it?

huningxin commented on services/ml/public/mojom/constants.mojom:
```diff
@@ -13,6 +13,8 @@ const int32 UINT32 = 2;
 const int32 TENSOR_INT32 = 4;
 const int32 TENSOR_QUANT8_ASYMM = 5;
+const int32 TENSOR_QUANT8_ASYMM_SIGNED = 14;
+const int32 TENSOR_QUANT8_SYMM_PER_CHANNEL = 11;
```
Could you please keep the order by constant value?

huningxin commented on third_party/blink/renderer/modules/ml/execution.cpp:
```diff
@@ -38,6 +38,10 @@ uint32_t requiredSize(int32_t type, const WTF::Vector<uint32_t>& dimensions) {
     return product(dimensions) * sizeof(int32_t);
   } else if (type == ml::mojom::blink::TENSOR_QUANT8_ASYMM) {
     return product(dimensions) * sizeof(int8_t);
```
It should be `uint8_t`, please help fix it.

huningxin commented on third_party/blink/renderer/modules/ml/model.cpp:
```diff
@@ -64,12 +64,22 @@ bool InvalidOperand(const OperandOptions* options,
       if (!options->hasDimensions())
         return InvalidState("Data type is invalid.", exception_state);
       break;
+    case NeuralNetworkContext::kTensorQuant8SymmPerChannel:
```
Could you please refer to https://github.com/intel/webml-polyfill/blob/master/src/nn/Model.js#L225 for validation?

lisa0314 replied:
Yes, I add the validation code into the `setOperandSymmPerChannelQuantParams` method.

huningxin commented on third_party/blink/renderer/modules/ml/neural_network_context.h:
```diff
@@ -34,6 +34,8 @@ class NeuralNetworkContext final : public ScriptWrappable,
   static const unsigned long kTensorInt32 = 4;
   static const unsigned long kTensorQuant8Asymm = 5;
+  static const unsigned long kTensorQuant8AsymmSigned = 14;
+  static const unsigned long kTensorQuant8SymmPerChannel = 11;
```
Please keep the order by constant value.

huningxin commented on third_party/blink/renderer/modules/ml/neural_network_context.idl:
```diff
@@ -10,6 +10,8 @@ interface NeuralNetworkContext {
     const long TENSOR_INT32 = 4;
     const long TENSOR_QUANT8_ASYMM = 5;
+    const long TENSOR_QUANT8_ASYMM_SIGNED = 14;
+    const long TENSOR_QUANT8_SYMM_PER_CHANNEL = 11;
```
ditto.

huningxin commented on third_party/blink/renderer/modules/ml/operand_symm_per_channel_quant_param.idl:
```diff
@@ -0,0 +1,10 @@
+
+dictionary OperandSymmPerChannelQuantParam {
+  unsigned long channelDim;
+  // required Float32Array scales;
```
If this comment is not used, please remove it.

huningxin reviewed (commented):
Thanks @lisa0314 , I've finished my review and left some comments. Please take a look.

lisa0314 commented:
@huningxin Thanks for your review! I have update the code as your suggestion and please help to review again!

huningxin commented on third_party/blink/renderer/modules/ml/model.cpp:
```diff
@@ -174,6 +192,36 @@ void Model::addOperation(int32_t type,
+  if (operandSymmPerChannelQuantParam->channelDim() < 0 ||
+      operandSymmPerChannelQuantParam->channelDim() >=
+          operand->dimensions.size())
+    return;
```
It seems to return silently. Would you like to indicate this is an invalid parameter?

lisa0314 replied:
I add a new method `InvalidOperandSymmPerChannelQuantParams` to indicate the invalid message.

huningxin commented on third_party/blink/renderer/modules/ml/model.cpp:
```diff
@@ -174,6 +192,36 @@ void Model::addOperation(int32_t type,
+
+  if (operandSymmPerChannelQuantParam->scales().size() !=
+      operand->dimensions[operandSymmPerChannelQuantParam->channelDim()])
+    return;
```
ditto

huningxin commented on third_party/blink/renderer/modules/ml/model.cpp:
```diff
@@ -174,6 +192,36 @@ void Model::addOperation(int32_t type,
+  for (uint32_t i = 0; i < operandSymmPerChannelQuantParam->scales().size();
+       i++) {
+    if (operandSymmPerChannelQuantParam->scales()[i] < 0) {
+      return;
```
ditto.

huningxin commented on services/ml/compilation_impl.cc:
```diff
@@ -209,10 +211,51 @@ int32_t CompilationImpl::GetConvParams(const mojom::OperationPtr& operation,
+      int32_t channel_dim = filter_param->channel_dim;
+      if ((channel_dim == 0 && !params.depthwise) ||
+          (channel_dim == 3 && params.depthwise)) {
+        params.scales_params.output_mask = 1 << 1;
```
I think the code that sets up the `DNNLInt8ScalesParams` should move to compilation_delegate_dnnl.cc.

lisa0314 replied:
I agree with you and move the related code to compilation_delegate_dnnl.cc.

huningxin commented on services/ml/compilation_impl.h:
```diff
@@ -57,6 +67,7 @@ struct ConvParams {
   uint32_t dilation_height;
   uint32_t depthwise_multiplier;
   int32_t fuse_code;
+  DNNLInt8ScalesParams scales_params;
```
If it is only used by compilation_delegate_dnnl, it should be moved to there.

huningxin reviewed (commented):
@lisa0314 , thanks for the update. It looks much better. I've left some comments for `DNNLInt8ScalesParams`. Please take a look.

huningxin commented:
@lisa0314 , any updates. Do you need me to do an other review? Thanks.

lisa0314 commented:
@huningxin , I has updated code as you previous comments. I only remain `input_scale` and `output_scale` in the `compilation_impl.h` and move the `weights_scales` and the calculation of output_scale into `compilation_delegate_dnnl`. I'm not sure that it's ok or not.

huningxin commented on services/ml/compilation_impl.h:
```diff
@@ -32,6 +32,12 @@ struct ElementWiseParams {
+  int32_t filter_type;
+  int32_t bias_type;
+  float input_scale;
+  float output_scale;
```
I'd recommend to remove these members as operand info already carries them.

huningxin commented on services/ml/compilation_impl.cc:
```diff
@@ -187,17 +187,22 @@ int32_t CompilationImpl::GetConvParams(const mojom::OperationPtr& operation,
   params.output_height = output->dimensions[1];
   params.output_width = output->dimensions[2];
   params.output_channel = output->dimensions[3];
+  params.output_scale = output->scale;
```
Please remove it here, dnnl delegate can handle that.

huningxin commented on services/ml/compilation_impl.cc:
```diff
@@ -187,17 +187,22 @@ int32_t CompilationImpl::GetConvParams(const mojom::OperationPtr& operation,
   params.input_height = input->dimensions[1];
   params.input_width = input->dimensions[2];
   params.input_channel = input->dimensions[3];
+  params.input_scale = input->scale;
```
ditto.

huningxin commented on services/ml/compilation_impl.cc:
```diff
@@ -209,10 +214,12 @@ int32_t CompilationImpl::GetConvParams(const mojom::OperationPtr& operation,
   const uint32_t bias_idx = inputs[index++];
   mojom::OperandPtr& bias = model_info_->operands[bias_idx];
   params.bias_length = bias->dimensions[0];
+  params.bias_type = bias->type;
```
ditto.

huningxin commented on services/ml/compilation_delegate_dnnl.h:
```diff
@@ -20,6 +20,12 @@ namespace ml {
+  int32_t output_mask;
+  uint32_t output_scales_count;
+  float* output_scales;
+};
```
It is only used in a local code block, do we really need to define it?

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -739,50 +745,59 @@ int32_t CompilationDelegateDnnl::AddConvolution(
   dnnl_status_t status;
   dnnl_memory_desc_t input_desc;
+  dnnl_data_type_t input_type;
+  GetDataType(params.input_type, &input_type);
```
Please use type of mojom::OperandPtr.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -739,50 +745,59 @@ int32_t CompilationDelegateDnnl::AddConvolution(
   }
   dnnl_memory_desc_t weights_desc;
+  dnnl_data_type_t weights_type;
+  GetDataType(params.filter_type, &weights_type);
```
ditto.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -739,50 +745,59 @@ int32_t CompilationDelegateDnnl::AddConvolution(
   }
 
+  dnnl_data_type_t bias_type;
+  GetDataType(params.bias_type, &bias_type);
```
ditto.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -739,50 +745,59 @@ int32_t CompilationDelegateDnnl::AddConvolution(
   }
   dnnl_memory_desc_t output_desc;
+  dnnl_data_type_t output_type;
+  GetDataType(params.output_type, &output_type);
```
ditto.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -814,9 +829,65 @@ int32_t CompilationDelegateDnnl::AddConvolution(
+  }
+
+  if (input_type == dnnl_s8 || input_type == dnnl_u8) {
+    DNNLInt8ScalesParams scales_params;
```
do we really need this structure?

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -814,9 +829,65 @@ int32_t CompilationDelegateDnnl::AddConvolution(
+          (float*)malloc(sizeof(float) * scales_params.output_scales_count);
+      for (uint32_t i = 0; i < scales_params.output_scales_count; i++) {
+        scales_params.output_scales[i] =
+            (params.input_scale * weights_scales[i]) / params.output_scale;
```
please use scale of mojom::OperandPtr directly.

huningxin commented on services/ml/compilation_delegate_dnnl.cc:
```diff
@@ -814,9 +829,65 @@ int32_t CompilationDelegateDnnl::AddConvolution(
+      scales_params.output_scales =
+          (float*)malloc(sizeof(float) * scales_params.output_scales_count);
+      scales_params.output_scales[0] =
+          (params.input_scale * weights_scale) / params.output_scale;
```
ditto.

huningxin reviewed (commented):
> I only remain input_scale and output_scale in the compilation_impl.h and move the weights_scales and the calculation of output_scale into compilation_delegate_dnnl.

I'd prefer to use scale of operand directly.

I did review and left some comments. Please take another look.

lisa0314 commented:
@huningxin thanks for your review and I have updated the code as you suggestion. I moved the scale and type into delegate_dnnl and use mojom::operandPtr directly.

huningxin reviewed (approved):
Looks good. Thanks for the PR @lisa0314 !
